OneSignal doc (#790)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/onesignal.mdx

</div>

- **Toggle on to create a device using email**: Enable this setting to create a device using email.
- **Toggle on to create a device using phone number**: Enable this setting to create a device using phone number.
- **Toggle on to concatenate event name with properties**: Enable this setting to concatenate the event names with properties. for example, if `add_to_cart` is an event and `brand` is a property, the event will be sent as `add_to_cart_brand` tag.
- **Allowed Property List**: Enter the properties you want to add as device tags, if present in the payload.

## Identify 

You can make an <Link to="/event-spec/standard-events/identify">`identify`</Link> call to add a new device to OneSignal App. However, if a device is already registered with the specified identifier, then it will update the existing device's record instead of creating a new one.

A sample `identify` call is shown below:

```javascript
rudderanalytics.identify('1hKOmRA4el9Zt1WSfVJIVo4GRlm', {
            firstName: 'Alex',
            lastName: 'Keener',
            email: "alex@example.com"
        }, {
            integrations: {
                oneSignal: {
                    deviceType: "6",

>                     identifier: ""

Reviewer: 7abcd558f29d0b1f048083e2834ad8ea4b3d87d8ad9c088b33c132706ff445f0

                }
            }
        );
```

You can create multiple devices in a single `identify` call:

> - One for the browser (by default)

Reviewer: We are not creating the device for browsers only. We are also adding device from the data of integrations object and depending on the context.device data. So it is also possible to add device of ios and android types. When we are setting deviceType as ios or android, we are setting device identifier as device.token, one can set this token for both [android](https://www.rudderstack.com/docs/sources/sdks/rudderstack-android-sdk/#setting-the-android-device-token) and [ios](https://www.rudderstack.com/docs/sources/sdks/rudderstack-ios-sdk/ios-v2/#setting-the-device-token-for-push-notifications). And if token is not available we are setting device.id as identifier. But it is recommended to send token only for push notifications.

| `traits.address.country` | `country` | Optional |
| `context.timezone` | `timezone` | Optional | 
| `anonymousId` | `external_user_id` | Optional | 

## Track

The <Link to="/event-spec/standard-events/track">`track`</Link> call can be used to update an existing device's tags in one of your OneSignal apps using the `external_user_id`. All devices containing the particular `external_user_id` are updated simultaneously.

While updating a device tag, any key inside a tag can be deleted by providing the key value as an empty string.

<div class="infoBlock">
Only the properties mentioned under the <strong>Allowed Property List</strong> setting in the RudderStack dashboard can be added as device tags.
</div>

A sample `track` call is shown below:

```javascript
rudderanalytics.track('Add to cart', {
    purchased_item: "Shirt",
		brand: "Zara"
	}
);
```

## Group

The <Link to="/event-spec/standard-events/group">`group`</Link> call can be used to group the devices having device tags using  the `groupId`.

A sample `group` call is as shown:

```javascript
rudderanalytics.group('1hKOmRA4el9Zt1WSfVJIVo4GRlm', {
    firstName: 'Alex',
    lastName: 'Keener',

>     email: "alex@example.com"

Reviewer: Updated.

});
```

### Property mapping

The following table details the mappings between RudderStack and OneSignal properties:

| RudderStack property | OneSignal property | Presence | 
| :-------| :--------| :------|
| `groupId` | `groupId` | Required | 
| `deviceType` | `device_type` | Required | 
| `traits.firstName` | `tags.first_name` | Optional |
| `traits.lastName` | `tags.last_name` | Optional | 
| `context.device.deviceModel` | `device_model` | Optional | 
| `context.os` | `device_os` | Optional | 
| `traits.address.country` | `country` | Optional |
| `context.timezone` | `timezone` | Optional | 

> | `anonymousId` | `external_user_id` | Optional | 

Reviewer: We are adding a groupId tag and setting the value as the value of groupId. Apart from that, we are adding all the group call traits to tags similar to track call properties.

Author: @ItsSudip I see these are being updated in the research spec, let me know when it's finalized. I'll pick these mappings up.

---
title: "OneSignal"
description: Step-by-step guide on sending your event data from RudderStack to OneSignal.
---

> [OneSignal](https://onesignal.com/) is a popular service which helps in driving customer engagement with multi-channel messaging across web and mobile push notifications, in-app messages, SMS, and email subscribers.

Reviewer 2: ```suggestion
[OneSignal](https://onesignal.com/) is a popular service that helps you drive customer engagement with multi-channel messaging across web and mobile push notifications, in-app messages, SMS, and email subscribers.
```

| **Connection Mode** | **Web** | **Mobile** | **Server** |
| :--- | :--- | :--- | :--- |
| **Device mode** | - | - | - |
| **Cloud mode** | **Supported**  | **Supported** | **Supported** |

<div class="infoBlock">
To know more about the difference between cloud mode and device mode in RudderStack, refer to the <Link to="/destinations/rudderstack-connection-modes/">RudderStack Connection Modes</Link> guide.
</div>

Once you have confirmed that the source platform supports sending events to OneSignal, follow these steps:

1. From your [RudderStack dashboard](https://app.rudderstack.com/), add a source. Then, from the list of destinations, select **OneSignal**.
2. Assign a name to the destination and click **Continue**.

### Connection settings

To successfully configure OneSignal as a destination, you will need to configure the following settings:

<img src="../../assets/event-stream-destinations/onesignal-connection-settings.png" alt="OneSignal connection settings"/>

- **App Id**: Enter your OneSignal App Id.

<div class="infoBlock">

> For more information on finding the OneSignal App Id, refer to the <a href="#faq">FAQ</a> section below.

Reviewer 2: ```suggestion
For more information on finding the OneSignal App Id, refer to the <Link to="#faq">FAQ</Link> section below.
```

</div>

> - **Toggle on to create a device using email**: Enable this setting to create a device using email.

Reviewer 2: @ujjwal-ab @ItsSudip can we rename this setting to `Add a device` in the web app to be consistent with the OneSignal feature? Reference: https://documentation.onesignal.com/reference/add-a-device 

cc @ishachopra0401

Reviewer 3: Ya, it's updated now in webapp. We can update in the doc too and can replace it with.
`Toggle on to add a device using email`
`Toggle on to add a device using phone number`

Author: @ujjwal-ab I dont see the changed settings in webapp. Can you please confirm?

Reviewer 3: Ya, actually, It is just changed in the master branch. That's why not visible in the develop. Doing it for develop also.

- **Toggle on to create a device using phone number**: Enable this setting to create a device using phone number.
- **Toggle on to concatenate event name with properties**: Enable this setting to concatenate the event names with properties. for example, if `add_to_cart` is an event and `brand` is a property, the event will be sent as `add_to_cart_brand` tag.

> - **Allowed Property List**: Enter the properties you want to add as device <Link to="https://documentation.onesignal.com/docs/add-user-data-tags">tags</Link>, if present in the payload. 

Reviewer 2: ```suggestion
- **Allowed Property List**: Enter the properties you want to add as the device <Link to="https://documentation.onesignal.com/docs/add-user-data-tags">tags</Link>, if present in the payload. 
```

<div class="warningBlock">

> The properties mentioned under <strong>Allowed Property List</strong> are considered only for the <code class="inline-code">track</code>  and <code class="inline-code">group</code> calls.

Reviewer 2: ```suggestion
RudderStack considers the properties mentioned under <strong>Allowed Property List</strong> only for the <code class="inline-code">track</code> and <code class="inline-code">group</code> calls.
```

</div>

## Identify 

> You can make an <Link to="/event-spec/standard-events/identify">`identify`</Link> call to add a new device to OneSignal App. However, if a device is already registered with the specified identifier, then it will update the existing device's record. Otherwise, it will create a new one.

Reviewer 2: ```suggestion
You can make an <Link to="/event-spec/standard-events/identify">`identify`</Link> call to add a new device to your OneSignal App. However, if a device is already registered with the specified identifier, then it will update the existing device's record.
```

Reviewer 2: Removing the last bit as we already say that in the first sentence.

A sample `identify` call is shown below: 

```javascript
rudderanalytics.identify('1hKOmRA4el9Zt1WSfVJIVo4GRlm', {
    firstName: 'Alex',
    lastName: 'Keener',
    email: "alex@example.com"
}, {
    externalId: [
        {
            type: "playerId",
            id: "Df344sdFgdDsS4"
        }
    ],
    integrations: {
        oneSignal: {
            deviceType: "6",
            identifier: "9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d"
        }
    }
);
```

In the above snippet, the `playerId` is a unique ID for a device. If `playerId` is provided in `externalId`, the device having that `playerId` will be updated. The `externalId` is case-sensitive and is expected to be in the above format for correct mapping.

> The `anonymoysId` is added as a tag with key as `anonymousId` and value as `anonymousId`'s value.

Reviewer 2: ```suggestion
<div class="infoBlock">
RudderStack adds the <code class="inline-code">anonymousId</code> as a tag with the key as <code class="inline-code">anonymousId</code> along with the corresponding value.
</div>
```

> You can create multiple devices in a single `identify` call:

Reviewer 2: @ishachopra0401 Should we have a separate sub-section for this?

> - For browser, RudderStack maps the `device_type` from the browser name and device type mapping as explained in the below code snippet. The `anonymousId` is set as an `identifier` depending upon the browser name. 

Reviewer 2: ```suggestion
- For the browser, RudderStack maps the `device_type` from the browser name and device type mapping as explained in the below code snippet. It sets the `anonymousId` as an `identifier` depending upon the browser name. 
```

>     RudderStack maps the following device types by default:

Reviewer 2: ```suggestion
RudderStack maps the following device types by default:
```

    ```javascript
    deviceTypeMapping = {
        android: 1,
        ios: 0,
        chrome: 5,
        safari: 7,
        firefox: 8
    }
    ```

>     You can override the device type mapping mentioned above or set any other device type by providing the `deviceType` and `identifier` in the `integrations` object, as shown:

Reviewer 2: ```suggestion
You can override the above-mentioned device type mapping or set any other device type by providing the `deviceType` and `identifier` in the `integrations` object, as shown:
```

    ```javascript
    "integrations": {
        "one_signal": {
            "deviceType": "Sample device",
            "identifier": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d"
        }
    }
    ```

> - For iOS and Android, the `device_type` is taken from the `context.device.type` and is mapped with the `device_type` value from the mappings mentioned above. The `context.device.token`/`context.device.id` is set as an `identifier`. 

Reviewer 2: ```suggestion
- For iOS and Android, RudderStack takes the `device_type` from the `context.device.type` field and maps it with the `device_type` value from the above-mentioned mappings. The `context.device.token`/`context.device.id` is set as an `identifier`. 
```

Reviewer 2: @ujjwal-ab @ItsSudip please confirm if it should be `deviceType` or `device_type`. cc @ishachopra0401

Reviewer 3: We are accepting `deviceType` in the integrations object and is mapped and send to OneSignal as `device_type`.

<div class="infoBlock">

>  It is recommended to send <code class="inline-code">device.token</code> for push notifications.

Reviewer 2: ```suggestion
It is recommended to send <code class="inline-code">device.token</code> for push notifications.
```

</div>

> - For email and SMS, enable the **Toggle on to create a device using email** and **Toggle on to create a device using phone number** settings in the RudderStack dashboard as well as send their data (`email` or `phone`) in the payload.

Reviewer 2: ```suggestion
- For email and SMS, enable the **Toggle on to create a device using email** and **Toggle on to create a device using phone number** settings in the RudderStack dashboard and send their data (`email` or `phone`) in the event payload.
```

> ### Property mapping

Reviewer 2: ```suggestion
### Supported mappings
```

> The following table details the mappings between RudderStack and OneSignal **optional** properties:

Reviewer 2: ```suggestion
The following table details the mappings between the **optional** RudderStack and OneSignal properties:
```

| RudderStack property | OneSignal property |
| :-------| :--------| 
| `context.device.model`  | `device_model` | 
| `context.os.version` | `device_os` | 
| `context.timezone` | `timezone` | 
| `userId` | `external_user_id` | 
| `context.locale` | `language` |
| `traits.createdAt`/`context.traits.createdAt`/`timestamp`/`originalTimestamp` | `created_at`/`last_active` | 
| `traits.country`/`context.traits.country`/`traits.address.country`/`context.traits.address.country` | `country` | 

<div class="infoBlock">
All the <code class="inline-code">string</code> key-value pairs from <code class="inline-code">traits</code> are mapped inside the <code class="inline-code">tags</code> object as it is.
</div>

## Track

> The <Link to="/event-spec/standard-events/track">`track`</Link> call can be used to update an existing device's tags in one of your OneSignal apps using the `external_user_id`. All devices containing the particular `external_user_id` are updated simultaneously.

Reviewer 2: ```suggestion
You can use the <Link to="/event-spec/standard-events/track">`track`</Link> to update an existing device's tags in your OneSignal apps using the `external_user_id` parameter. All devices containing the particular `external_user_id` are updated simultaneously.
```

> While updating a device tag, any key inside a tag can be deleted by providing the key value as an empty string.

Reviewer 2: ```suggestion
While updating a device tag, you can delete any key inside a tag by providing the key value as an empty string.
```

A sample `track` call is shown below:

```javascript
rudderanalytics.track('Add to cart', {
    purchased_item: "Shirt",
		brand: "Zara"
	}
);

> ```

Reviewer 2: ```

rudderanalytics.track('Add to cart', {
  purchased_item: "Shirt",
  brand: "Zara"
});
```

Can we use the above beautified code?

### Property mapping

The following table details the mappings between RudderStack and OneSignal properties:

| RudderStack property | OneSignal property | Presence | 
| :-------| :--------| :------|
| `userId`  | `external_user_id` | Required | 

<div class="infoBlock">
All the properties added in the <strong>Allowed Property List</strong> setting in the RudderStack dashboard will be mapped inside the <code class="inline-code">tags</code> object if available in the payload. However, the key-value pair should be of data type <code class="inline-code">string</code> only.
</div>

## Group

> The <Link to="/event-spec/standard-events/group">`group`</Link> call can be used to group the devices having device tags using  the `groupId`.

Reviewer 2: ```suggestion
You can use the <Link to="/event-spec/standard-events/group">`group`</Link> call to group the devices with the device tags using the same `groupId`.
```

Reviewer 2: @ujjwal-ab @ItsSudip @ishachopra0401 please confirm if the above change is technically sound.

Reviewer 3: LGTM

A sample `group` call is as shown:

```javascript
rudderanalytics.group('1hKOmRA4el9Zt1WSfVJIVo4GRlm', {
  name: "Apple Inc.",
  location: "USA",
});
```

### Property mapping

The following table details the mappings between RudderStack and OneSignal properties:

| RudderStack property | OneSignal property | Presence | 
| :-------| :--------| :------|
| `groupId` | `groupId` | Required | 

<div class="infoBlock">

> All the properties added in the <strong>Allowed Property List</strong> setting in the RudderStack dashboard will be mapped inside the <code class="inline-code">tags</code> object if available in the payload. However, the key value pair should be of data type <code class="inline-code">string</code> only.

Reviewer 2: ```suggestion
RudderStack maps all the properties added in the <strong>Allowed Property List</strong> setting in the RudderStack dashboard to the <code class="inline-code">tags</code> object if they are present in the payload. However, note that the key-value pair should be of the String data type only.
```
